### PR TITLE
DEV-4562 | Solve for importing pageless contributions in Stripe transactions importer

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

The base for this PR is initially the DEV-4659 branch. This is because this PR's branch (DEV-4562) branches from DEV-4659. 

Once DEV-4569 has been merged into main, the base for this PR should be changed to main.

#### What's this PR do?

This PR alters our stripe transactions import code such that we can import "pageless" Stripe transactions. In previous import work, it arose that some relevant Stripe payment intents and subscriptions will not have a value for referer, and we use this value to link a transaction to a donation page.  This linkage is critical because it by transitive property links a contribution to a Stripe account, without which different parts of the app break (for instance, updating/modifying subscriptions).

To solve for this problem, we considered 4 different possible solutions before deciding on the solution implemented in this PR ([see this confluence page for more info](https://news-revenue-hub.atlassian.net/wiki/spaces/TECH/pages/2778333190/DEV-4562+Model+changes+to+ensure+referential+integrity+for+stripe+account+ID+on+imported+contributions)). The solution we chose is to add a new foreign key to our Contribution model to the Revenue Program model.  Additionally, we add a constraint (using Django's `CheckConstraint`) that requires that a contribution have either an FK to a page (preferred) or an FK to a revenue program, but not both.

**Note**: This iteration of the import transactions command will take longer than the version in DEV-4659 insofar as pageless contributions now get processed, whereas before they were excluded.

The following are some more specific changes:

- Changes to `apps.contributions.admin.ContributionAdmin`
    - Updates admin detail view to display the `._revenue_program` field, and provides description explaining in admin that only DP or RP can be set, and that there is a separate revenue program property that can be found in a different section of the admin page.
    - Removes the ability to create and update Contributions from the Django admin. There is no use case for either of these. In status quo, creation is enabled, but is practically impossible because `amount` is a read only field, and a contribution have an amount. The only field that can be updated in status quo is `contribution_metadata` and there is no use case for editing that directly in Django admin (note these changes were approved by Daniel in a Slack thread).
- Updates the `sync_payment_transaction_time` command to account for new data modeling.
-  Updates `apps.contributions.filters.PortalContributionFilter` to also consider contributions connected by `._revenue_program`d
- Changes to the `Contribution` model:
    - Adds `._revenue_program` as a new field. This is an FK to a revenue program.
    - Adds a constraint to model that requires that either `donation_page` or `._revenue_program` (but not both) be set.
    - Updates `Contribution.revenue_program` property to return an RP via donation page and barring that from the `._revenue_program` field (which is where FK to RP is stored).
- Updates to `apps.contributions.stripe_import.py` to account for new data modeling:
    - When attempting to import a transaction, we prefer to link it to a donation page, but if none can be found, we try to link it to an RP.
    - Removes `StripeTransactionsImporter.conditionally_update_donation_page` command.  In reimplemented version, we pass either page or RP as defaults to the upsert call, and don't do a subsequent update.
    - Adds `_revenue_program` and `donation_page` as `dont_update` fields in the call to `upsert_with_diff_check`. If either DP or RP is set on an existing contribution and we're upserting, we don't want to override either existing value.
- Adds a test for the `process_stripe_event` management command which appears to have been missed in earlier PRs. We already tested the class that gets used in the command, but missed testing the command itself.
- Changes many references to `contribution.donation_page.revenue_program` to instead use `contribution.revenue_program` so will work regardless of source of RP value.

#### Why are we doing this? How does it help us?

- Enables importing legacy contributions
- In turn that unblocks releasing new portal

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Coming soon

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

[Confluence page](https://news-revenue-hub.atlassian.net/wiki/spaces/TECH/pages/2778333190/DEV-4562+Model+changes+to+ensure+referential+integrity+for+stripe+account+ID+on+imported+contributions)

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-4562](https://news-revenue-hub.atlassian.net/browse/DEV-4562)
[DEV-4751](https://news-revenue-hub.atlassian.net/browse/DEV-4751)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

Yes. Follow up ticket to fix bug around provider subscription ID being null for created recurring contributions, discovered while testing.

https://news-revenue-hub.atlassian.net/browse/DEV-4751

[DEV-4562]: https://news-revenue-hub.atlassian.net/browse/DEV-4562?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/models.py [hunk 1]
```diff
@@ -137,7 +137,8 @@ def filtered_by_role_assignment(self, role_assignment: RoleAssignment) -> models
                 )
```
Comment (njharman) [inline apps/contributions/models.py:137]: line #136   `donation_page__revenue_program__organization=role_assignment.organization`

Does this require dual dp/_rp treatment? If not, maybe short comment to explain the difference with RP_ADMIN query.

**Review by njharman — CHANGES_REQUESTED** [review]

File: apps/contributions/models.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:138]: Great catch. Thanks!

bf908920b1

File: apps/contributions/models.py [hunk 2]
```diff
@@ -234,6 +258,7 @@ def revenue_program(self) -> RevenueProgram | None:
         # TODO: [DEV-4507] Remove this property and replace with a direct FK to RevenueProgram
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: No longer relevant given model changes. 2e16c67f46

File: apps/contributions/stripe_import.py [hunk 3]
```diff
@@ -784,23 +789,15 @@ def upsert_contribution(self, stripe_entity: dict, is_one_time: bool) -> Tuple[C
             # If there's contribution metadata, we want to leave it intact.
             # Otherwise we see spurious updates because of key ordering in the
             # metadata and conversions of null <-> None.
-            dont_update=["contribution_metadata"],
-        )
-        updated_contribution = self.conditionally_update_contribution_donation_page(
-            contribution=contribution, donation_page=self.get_donation_page_from_metadata(metadata)
+            # additional note
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/stripe_import.py]: That was a note to myself to leave a comment, which I didn't get back to. Fixed in 84af7f4ec

**Review by x110dc — APPROVED** [review]

**Pull request merged**